#' ORFtrap_call_hits
#'
#' @description Function to call hits from a given screen. Receives sorted and unsorted counts as input, computes FC table
#'
#' @param sorted.forward.counts Character vector of sorted (FACSed) count file paths generated by vl_ORFtag_pipeline().
#' Counts from all provided files will be merged.
#' @param unsorted.forward.counts Character vector of unsorted (input) count file paths generated by vl_ORFtag_pipeline().
#' Counts from all provided files will be merged.
#' @param genome Genome annotation to be used.
#' For now, only "mm10" and "hg38" are supported, and the corresponding non-first-exon .gtf will be used for sorting.
#' See at the bottom of the ?vl_ORFtag_pipeline help page to see how to generate custom non-first-exon start .gtf files for other genomes.
#' @param name Name prefix appended to output file names.
#' @param padj.cutoff The p.adjust cutoff to be used to call hits (>=). Default= 0.001
#' @param log2OR.cutoff The log2OR cutoff to be used to call hits (>=). Default= 1
#' @param log2OR.pseudocount The pseudocount to be added to avoid infinite values. Note that only the log2OR will be affected, not the fisher p.values. Default= 1
#' @param output.suffix Name suffix appended to output file names. Default= "_vs_unsorted.txt".
#' @param output.folder.FC.file Output folder for FC files. Default= "db/FC_tables/ORFtag".
#'
#' @return Returns FC tables containing DESeq2-like columns.
#'
#' @examples
#' # Call hits ----
#' # 124 hits should be identified with this dataset
#' callOrftagHits(sorted.forward.counts = c("db/exon_assignment/ORFtag/Activator2_sort_rep1_same_strand.txt",
#'                                                "db/exon_assignment/ORFtag/Activator2_sort_rep2_same_strand.txt"),
#'                      unsorted.forward.counts = c("db/exon_assignment/ORFtag/Activator2_input_rep1_same_strand.txt",
#'                                                  "db/exon_assignment/ORFtag/Activator2_input_rep2_same_strand.txt"),
#'                      genome = "mm10",
#'                      name = "Activator_2",
#'                      output.suffix = "_vs_input.txt")
#'
#' # Call hits using revese strand (sanity check -> be cautious with the hits that are also found here!) ----
#' # 81 hits should be identified with the reverse strand
#' callOrftagHits(sorted.forward.counts = c("db/exon_assignment/ORFtag/Activator2_sort_rep1_rev_strand.txt",
#'                                                "db/exon_assignment/ORFtag/Activator2_sort_rep2_rev_strand.txt"),
#'                      unsorted.forward.counts = c("db/exon_assignment/ORFtag/Activator2_input_rep1_rev_strand.txt",
#'                                                  "db/exon_assignment/ORFtag/Activator2_input_rep2_rev_strand.txt"),
#'                      genome = "mm10",
#'                      name = "Activator_2",
#'                      output.suffix = "_vs_input_rev_strand.txt")
#'
#' # 4 of the hits should be identified with the reverse strand and should be considered carefully
#' hits <- fread("db/FC_tables/ORFtag/Activator_2_vs_input.txt")[(hit), gene_name]
#' sanityCheck <- fread("db/FC_tables/ORFtag/Activator_2_vs_input_rev_strand.txt")[(hit), gene_name]
#'
#' # Hits to be considered carefully:
#' intersect(hits, sanityCheck)
#'
#' # To call hits using strand bias (not used)
#' see ?callOrftagHits_strandBias()
#'
#' @export
callOrftagHits <- function(sorted.forward.counts,
                           unsorted.forward.counts,
                           genome,
                           name,
                           padj.cutoff= 0.001,
                           log2OR.cutoff= 1,
                           log2OR.pseudocount= 1,
                           output.suffix= "_vs_unsorted.txt",
                           output.folder.FC.file= "db/FC_tables/ORFtag/")
{
  require(rtracklayer)
  require(data.table)
  require(GenomicRanges)

  # Checks
  if(!is.character(name) | length(name)!=1)
    stop("name should be a unique character specifying the name of the screen")
  if(anyDuplicated(sorted.forward.counts))
    stop("Duplicated filenames in sorted.forward.counts")
  if(anyDuplicated(unsorted.forward.counts))
    stop("Duplicated filenames in unsorted.forward.counts")
  if(!is.character(output.suffix) | length(output.suffix)!=1)
    stop("output.suffix should be a unique character specifying the name of the screen")

  # Import counts
  input <- rbindlist(lapply(unsorted.forward.counts, fread))
  sample <- rbindlist(lapply(sorted.forward.counts, fread))
  dat <- rbindlist(list(count.input= input, count.sample= sample), idcol= "cdition")
  dat <- na.omit(dat[dist<2e5])
  total <- dat[, .N, .(cdition)]
  dat <- dat[, .(count= .N), .(gene_id, cdition)]
  dat <- dcast(dat, gene_id~cdition, value.var = "count")

  # Import gene exons
  gtf <- switch(genome,
                "mm10" = "/groups/stark/vloubiere/projects/ORFTRAP_1/db/gtf/exons_start_mm10.gtf",
                "hg38" = "/groups/stark/vloubiere/projects/ORFTRAP_1/db/gtf/exons_start_hg38.gtf")
  genes <- rtracklayer::import(gtf)
  genes <- as.data.table(mcols(genes)[, c("gene_id", "gene_name")])
  genes <- unique(genes)
  setorderv(genes, "gene_id")

  # Format dat
  dat <- merge(genes, dat, by= "gene_id", all.x= T)
  dat[is.na(count.input), count.input:= 0]
  dat[is.na(count.sample), count.sample:= 0]
  dat[, total.input:= total[cdition=="count.input", N]]
  dat[, total.sample:= total[cdition=="count.sample", N]]

  # Fisher test sample vs input
  dat[count.sample>=3, c("OR", "pval"):= {
    .t <- matrix(c(total.input-count.input,
                   count.input,
                   total.sample-count.sample,
                   count.sample),
                 ncol= 2)
    .(fisher.test(.t+log2OR.pseudocount, alternative = "greater")$estimate,
      fisher.test(.t, alternative = "greater")$p.value)
  }, .(count.input, total.input, count.sample, total.sample)]
  dat[count.sample>=3, log2OR:= log2(OR)]
  dat[count.sample>=3, padj:= p.adjust(pval, "fdr")]
  dat[, hit:= padj<=padj.cutoff & log2OR>=log2OR.cutoff]

  # Clean and save
  dat <- genes[dat, on= c("gene_id", "gene_name")]
  dat$OR <- dat$pval <- NULL
  dir.create(output.folder.FC.file, showWarnings = F, recursive = T)
  FC_table <- paste0(output.folder.FC.file, "/", name, output.suffix)
  fwrite(dat,
         FC_table,
         col.names = T,
         row.names = F,
         sep= "\t",
         quote= F,
         na= NA)
  cat(paste0(name, ": ", sum(dat$hit, na.rm = T), " hits were called!\nFC file -> ", FC_table, "\n"))
}
