#' callORFtagHitsStrandBias
#'
#' @description Function to call hits from a given screen. Receives sorted and unsorted counts as input, computes FC table
#'
#' @param sorted.forward.counts Character vector of sorted (FACSed) count file paths generated by vl_ORFtag_pipeline().
#' Counts from all provided files will be merged.
#' @param sorted.reverse.counts Character vector of sorted (FACSed), REVERSED count file paths generated by vl_ORFtag_pipeline().
#' Counts from all provided files will be merged.
#' @param unsorted.forward.counts Character vector of unsorted (input) count file paths generated by vl_ORFtag_pipeline().
#' Counts from all provided files will be merged.
#' @param unsorted.reverse.counts Character vector of insorted (input), REVERSED count file paths generated by vl_ORFtag_pipeline().
#' Counts from all provided files will be merged.
#' @param genome Genome annotation to be used. For now, only "mm10" and "hg38" are supported, and the corresponding non-first-exon .gtf will be used for sorting.
#' See at the bottom of the ?vl_ORFtag_pipeline help page to see how to generate custom non-first-exon start .gtf files for other genomes.
#' @param name Name prefix appended to output file names.
#' @param padj.cutoff The p.adjust cutoff to be used to call hits (>=). Default= 0.05.
#' @param log2OR.cutoff The log2OR cutoff to be used to call hits (>=). Default= 1.
#' @param log2OR.pseudocount The pseudocount to be added to avoid infinite values.
#' Note that only the log2OR will be affected, not the fisher p.values. Default= 1
#' @param binom.pseudocount The pseudocount to be added before binomial test (does not tolerate 0s). Default= 1
#' @param binom.padj.cutoff The p.adjust cutoff to be used to call binom_hits (<=). Default= 0.001.
#' @param output.suffix Name suffix appended to output file names. Default= "_vs_unsorted.txt".
#' @param output.folder.FC.file Output folder for FC files. Default= "db/FC_tables/ORFtag".
#'
#' @return Returns FC tables containing DESeq2-like columns.
#'
#' @examples
#' library(vlfunctions)
#'
#' # 63 hits should be called with this dataset ----
#' callORFtagHitsStrandBias(sorted.forward.counts = c("db/exon_assignment/ORFtag/Activator2_sort_rep1_same_strand.txt",
#'                                                           "db/exon_assignment/ORFtag/Activator2_sort_rep2_same_strand.txt"),
#'                                 sorted.reverse.counts = c("db/exon_assignment/ORFtag/Activator2_sort_rep1_rev_strand.txt",
#'                                                           "db/exon_assignment/ORFtag/Activator2_sort_rep2_rev_strand.txt"),
#'                                 unsorted.forward.counts = c("db/exon_assignment/ORFtag/Activator2_input_rep1_same_strand.txt",
#'                                                             "db/exon_assignment/ORFtag/Activator2_input_rep2_same_strand.txt"),
#'                                 unsorted.reverse.counts = c("db/exon_assignment/ORFtag/Activator2_input_rep1_rev_strand.txt",
#'                                                             "db/exon_assignment/ORFtag/Activator2_input_rep2_rev_strand.txt"),
#'                                 genome = "mm10",
#'                                 name = "Activator_2",
#'                                 output.suffix = "_vs_input_strandBias.txt")
#'
#' @export
callORFtagHitsStrandBias <- function(sorted.forward.counts,
                                     sorted.reverse.counts,
                                     unsorted.forward.counts,
                                     unsorted.reverse.counts,
                                     genome,
                                     name,
                                     padj.cutoff= 0.05,
                                     log2OR.cutoff= 1,
                                     log2OR.pseudocount= 1,
                                     binom.pseudocount= 1,
                                     binom.padj.cutoff= 0.001,
                                     output.suffix= "_vs_revStrand",
                                     output.folder.FC.file= "db/FC_tables/ORFtag")
{
  require(rtracklayer)
  require(data.table)
  require(GenomicRanges)

  # Checks
  if(!is.character(name) | length(name)!=1)
    stop("name should be a unique character specifying the name of the screen")
  if(anyDuplicated(sorted.forward.counts))
    stop("Duplicated filenames in sorted.forward.counts")
  if(anyDuplicated(unsorted.forward.counts))
    stop("Duplicated filenames in unsorted.forward.counts")
  if(!is.character(output.suffix) | length(output.suffix)!=1)
    stop("output.suffix should be a unique character specifying the name of the screen")

  # Import counts
  inputFw <- rbindlist(lapply(unsorted.forward.counts, fread))
  inputRev <- rbindlist(lapply(unsorted.reverse.counts, fread))
  sampleFw <- rbindlist(lapply(sorted.forward.counts, fread))
  sampleRev <- rbindlist(lapply(sorted.reverse.counts, fread))

  dat <- rbindlist(list(count.sample.fw= sampleFw,
                        count.sample.rev= sampleRev,
                        count.input.fw= inputFw,
                        count.input.rev= inputRev),
                   idcol= "cdition")
  dat <- na.omit(dat[dist<5e4])
  dat <- dat[, .(count= .N), .(gene_id, cdition)]
  dat <- dcast(dat,
               gene_id~cdition,
               value.var = "count",
               fill= 0)

  # Import gene exons
  gtf <- switch(genome,
                "mm10" = "/groups/stark/vloubiere/projects/ORFTRAP_1/db/gtf/exons_start_mm10.gtf",
                "hg38" = "/groups/stark/vloubiere/projects/ORFTRAP_1/db/gtf/exons_start_hg38.gtf")
  genes <- rtracklayer::import(gtf)
  genes <- as.data.table(mcols(genes)[, c("gene_id", "gene_name")])
  genes <- unique(genes)
  setorderv(genes, "gene_id")

  # Format dat
  dat <- merge(genes, dat, by= "gene_id", all.x= T)

  # Fisher test fw vs rev
  dat[count.sample.fw>=3, c("OR", "pval"):= {
    .t <- matrix(c(count.sample.fw,
                   count.sample.rev,
                   count.input.fw,
                   count.input.rev),
                 ncol= 2)
    .(fisher.test(.t, alternative = "greater")$estimate,
      fisher.test(.t+log2OR.pseudocount, alternative = "greater")$p.value)
  }, .(count.sample.fw, count.sample.rev, count.input.fw, count.input.rev)]
  dat[count.sample.fw>=3, log2OR:= log2(OR)]
  dat[count.sample.fw>=3, padj:= p.adjust(pval, "fdr")]
  dat[, hit:= padj<padj.cutoff & log2OR>=log2OR.cutoff]

  # Add binomial test
  dat[count.sample.fw>=3, binom_pval:= {
    binom.test(x = count.sample.fw+binom.pseudocount,
               n = count.sample.fw+count.sample.rev+binom.pseudocount,
               p = (count.input.fw+binom.pseudocount)/(count.input.fw+count.input.rev+binom.pseudocount),
               alternative = "greater")["p.value"]
  }, .(count.sample.fw, count.sample.rev, count.input.fw, count.input.rev)]
  dat[count.sample.fw>=3, binom_padj:= p.adjust(binom_pval, "fdr")]
  dat[, binom_hit:= binom_padj<binom.padj.cutoff]

  # Clean and save
  dat <- genes[dat, on= c("gene_id", "gene_name")]
  dat$OR <- dat$pval <- dat$binom_pval <- NULL
  dir.create(output.folder.FC.file, showWarnings = F, recursive = T)
  FC_table <- paste0(output.folder.FC.file, "/", name, output.suffix)
  fwrite(dat,
         FC_table,
         col.names = T,
         row.names = F,
         sep= "\t",
         quote= F,
         na= NA)
  cat(paste0(name, ": ", sum(dat$hit, na.rm = T), " hits were called!\nFC file -> ", FC_table, "\n"))
}
